### ISSUE TITLE
No code implementation about discriminative loss ?

### ISSUE THREAD

**Issue by LiZhenLiangLee** [speedinghzl/CCNet#104]
Section 3.4 in your TPAMI version paper,  **_Learning Category Consistent Features_**
 Is there any code for this part ?

**Comment by wqhIris** [comment]
I was also wondering this,
Is it possible for providing the implementation of this discriminative loss?

**Comment by ShinoSpace** [comment]
@LiZhenLiangLee @wqhIris Here's my guess.
The original paper published in TPAMI seems not strict enough to illustrate the implement detail. In fact, from this code, i find this model is a encoder-only model rather than auto-encoder like UNet. The upsample is simple bilinear interpolation, and the upsampled feature is the final segmentation result. Therefore, this main problem to implement the loss function proposed in **_Learning Category Consistent Features_** section is there's no feature embedding to "pull" and "push"(i.e. the loss component $L_var$ and $L_dis$).
Here's what i have tried.
I tried to re-implement this metric loss by adding bilinear interpolation followed by a 1x1 conv to the ccnet, while also code the loss function to optimize the feature(embedding) output from 1x1 conv. In fact, i found this loss is not easy to implement, and the training loss of my task seems a little unstable(but actually in convergence). I don't know whether the author of this repo also noticed this problem.
